src/Gui/DlgKeyboardImp.cpp

```diff
@@ -71,6 +77,231 @@ DlgCustomKeyboardImp::DlgCustomKeyboardImp( QWidget* parent  )
+            populateCommandGroups(combo);
+            combo->activated(combo->currentIndex());
+        }
+    });
+}
+
+void DlgCustomKeyboardImp::initPriorityList(QTreeWidget *priorityList,
+                                            QAbstractButton *buttonUp,
+                                            QAbstractButton *buttonDown)
+{
+    QStringList labels;
+    labels << tr("Name") << tr("Title");
+    priorityList->setHeaderLabels(labels);
+    priorityList->header()->hide();
+    priorityList->header()->setSectionResizeMode(0, QHeaderView::ResizeToContents);
+    priorityList->header()->setSectionResizeMode(1, QHeaderView::ResizeToContents);
+
+    auto updatePriorityList = [priorityList](bool up) {
+        auto item = priorityList->currentItem();
+        if (!item)
+            return;
+
+        int index = priorityList->indexOfTopLevelItem(item);
+        if (index < 0)
+            return;
+        if ((index == 0 && up)
+                || (index == priorityList->topLevelItemCount()-1 && !up))
+            return;
+
+        std::vector<QByteArray> actions;
+        for (int i=0; i<priorityList->topLevelItemCount(); ++i) {
+            auto item = priorityList->topLevelItem(i);
+            actions.push_back(item->data(0, Qt::UserRole).toByteArray());
+        }
+
+        auto it = actions.begin() + index;
+        auto itNext = up ? it - 1 : it + 1;
+        std::swap(*it, *itNext);
+        ShortcutManager::instance()->setPriorities(actions);
+    };
+
+    QObject::connect(buttonUp, &QAbstractButton::clicked, [=](){updatePriorityList(true);});
+    QObject::connect(buttonDown, &QAbstractButton::clicked, [=](){updatePriorityList(false);});
+    QObject::connect(priorityList, &QTreeWidget::currentItemChanged,
+        [=](QTreeWidgetItem *item){
+            buttonUp->setEnabled(item!=nullptr);
+            buttonDown->setEnabled(item!=nullptr);
+        }
+    );
+}
+
+boost::signals2::connection
+DlgCustomKeyboardImp::initCommandWidgets(QTreeWidget *commandTreeWidget,
+                                         QComboBox *comboGroups,
+                                         QLineEdit *editCommand,
+                                         QTreeWidget *priorityList,
+                                         QAbstractButton *buttonUp,
+                                         QAbstractButton *buttonDown,
+                                         AccelLineEdit *editShortcut,
+                                         AccelLineEdit *currentShortcut)
+{
+    initCommandCompleter(editCommand, comboGroups, commandTreeWidget);
+    auto conn = initCommandList(commandTreeWidget, comboGroups);
+
+    if (priorityList && buttonUp && buttonDown) {
+        initPriorityList(priorityList, buttonUp, buttonDown);
+
+        auto timer = new QTimer(priorityList);
+        timer->setSingleShot(true);
+        if (currentShortcut)
+            QObject::connect(currentShortcut, &QLineEdit::textChanged, timer, [timer](){timer->start(200);});

```

Comment:
How did you arrive at the 200ms for these?

Reply:
Just to prevent refreshing priority list too often, since it gets triggered in a few signals.